File: LogController.m
//  Converted to Swift 5.7 by Swiftify v5.7.28606 - https://swiftify.com/
//
//  LogController.m
//  Get_iPlayer GUI
//
//  Created by Thomas Willson on 7/9/14.
//
//

//
//  LogController.h
//  Get_iPlayer GUI
//
//  Created by Thomas Willson on 7/9/14.
//
//

import Cocoa
import CocoaLumberjack

class LogController: NSObject, DDLogFormatter {
    @IBOutlet var log: NSTextView!
    @IBOutlet weak var window: NSWindow!

    override init() {
        //Initialize Log
        super.init()
        let fileLogger = DDFileLogger()
        fileLogger.rollingFrequency = 60 * 60 * 24
        fileLogger.logFileManager.maximumNumberOfLogFiles = 1
        fileLogger.logFormatter = self
        DDLog.add(fileLogger)
        DDLog.add(DDOSLogger())
        var version: String? = nil
        if let object = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") {
            version = "\(object)"
        }
        DDLogInfo("Get iPlayer Automator %@ Initialized.", version)
    }

    override func awakeFromNib() {
        log.textColor = .white
        log.font = .userFixedPitchFont(ofSize: 12.0)
    }

    @IBAction func showLog(_ sender: Any) {
        window.makeKeyAndOrderFront(self)
        log.scrollToEndOfDocument(self)
    }

    @IBAction func copyLog(_ sender: Any) {
        let unattributedLog = log.string
        let pb = NSPasteboard.general
        let types = [.string] as? [AnyHashable]
        if let types {
            pb.declareTypes(types, owner: self)
        }
        pb.setString(unattributedLog, forType: .string)
    }

    @IBAction func clearLog(_ sender: Any) {
        log.string = ""
    }

    func formatLogMessage(_ logMessage: DDLogMessage?) -> String? {
        DispatchQueue.main.async(execute: { [self] in
            // In normal mode don't dump debug or verbose messages to the console.
            let verbose = UserDefaults.standard.bool(forKey: "Verbose")
            if !verbose && ((logMessage?.flag == DDLogFlagDebug) || (logMessage?.flag == DDLogFlagVerbose)) {
                return
            }

            let messageWithNewline = (logMessage?.message ?? "") + "\r"
            let newMessage = NSMutableAttributedString(string: messageWithNewline)

            var textColor = log.textColor

            switch logMessage?.flag {
            case DDLogFlagWarning:
                textColor = .yellow
            case DDLogFlagError:
                textColor = .red
            case DDLogFlagDebug:
                textColor = .lightGray
            case DDLogFlagVerbose:
                textColor = .gray
            default:
                // use base color.
                break
            }

            newMessage.addAttribute(
                .foregroundColor,
                value: textColor,
                range: NSRange(location: 0, length: newMessage.length))

            newMessage.addAttribute(
                .font,
                value: log.font,
                range: NSRange(location: 0, length: newMessage.length))

            log.textStorage?.append(newMessage)

            //Scroll log to bottom only if it is visible.
            if window.isVisible {
                let shouldAutoScroll = Int(log.bounds.maxY) == Int(log.visibleRect.maxY)
                if shouldAutoScroll {
                    log.scrollToEndOfDocument(nil)
                }
            }
        })

        return logMessage?.init()
    }
}

File: LogController.swift
//
//  LogController.swift
//  Get iPlayer Automator
//
//  Created by Scott Kovatch on 1/1/24.
//

import Foundation
import CocoaLumberjackSwift

@objc(LogController)
class LogController: NSObject, DDLogFormatter {
    @IBOutlet var log: NSTextView!
    @IBOutlet var window: NSWindow!

    override init() {
        super.init()

        let fileLogger = DDFileLogger()
        fileLogger.rollingFrequency = 60 * 60 * 24
        fileLogger.logFileManager.maximumNumberOfLogFiles = 1
        fileLogger.logFormatter = self
        DDLog.add(fileLogger)
        DDLog.add(DDOSLogger())
        if let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") {
            DDLogInfo("Get iPlayer Automator \(version) Initialized.")
        }
    }

    override func awakeFromNib() {
        log.textColor = NSColor.white
        log.font = NSFont.userFixedPitchFont(ofSize: 12)
    }

    @IBAction func showLog(sender: AnyObject) {
        window.makeKeyAndOrderFront(self)
        log.scrollToEndOfDocument(self)
    }

    @IBAction func copyLog(sender: AnyObject) {
        let unattributedLog = log.string
        let pb = NSPasteboard.general
        pb.declareTypes([.string], owner: self)
        pb.setString(String(unattributedLog), forType: .string)
    }

    @IBAction func clearLog(sender: AnyObject) {
        log.string = ""
    }

    func format(message: DDLogMessage) -> String? {
        // In normal mode don't dump debug or verbose messages to the console.
        let verbose = UserDefaults.standard.bool(forKey:"Verbose")
        if !verbose && ((message.flag == .debug) || (message.flag == .verbose)) {
            return nil
        }

        DispatchQueue.main.async {
            let messageWithNewline = message.message + "\r"
            var textColor = self.log.textColor ?? .white

            let newMessage = NSMutableAttributedString(string: messageWithNewline)
            switch message.flag {
            case .warning:
                textColor = NSColor.yellow
            case .error:
                textColor = NSColor.red
            case .debug:
                textColor = NSColor.lightGray
            case .verbose:
                textColor = NSColor.gray
            default:
                // use base color.
                break
            }

            newMessage.addAttribute(NSAttributedString.Key.foregroundColor,
                                    value:textColor,
                                    range:NSMakeRange(0, newMessage.length))

            if let font = self.log.font {
                newMessage.addAttribute(NSAttributedString.Key.font,
                                        value:font,
                                        range:NSMakeRange(0, newMessage.length))
            }
            self.log.textStorage?.append(newMessage)

            //Scroll log to bottom only if it is visible.
            if (self.window.isVisible) {
                let shouldAutoScroll = (NSMaxY(self.log.bounds) == NSMaxY(self.log.visibleRect))
                if (shouldAutoScroll) {
                    self.log.scrollToEndOfDocument(nil)
                }
            }
        }

        return message.message
    }
}
